src/mame/tandberg/tdv2100_disp_logic.cpp

```diff
@@ -0,0 +1,1029 @@
+	m_sw_invert_video(*this, "sw_invert_video"),
+	m_sw_rs232_baud(*this, "sw_rs232_baud"),
+	m_sw_rs232_settings(*this, "sw_rs232_settings"),
+	m_dsw_u39(*this, "dsw_u39"),
+	m_dsw_u73(*this, "dsw_u73"),
+	m_dsw_u61(*this, "dsw_u61"),
+	m_write_waitl_cb(*this),
+	m_write_onlil_cb(*this),
+	m_write_carl_cb(*this),
+	m_write_errorl_cb(*this),
+	m_write_enql_cb(*this),
+	m_write_ackl_cb(*this),
+	m_write_nakl_cb(*this),
+	m_vblank_state(false),
+	m_speed_check(false),
+	m_rx_handshake(false),
+	m_tx_handshake(false),
+	m_line_key_held(false),
+	m_trans_key_held(false),
+	m_break_key_held(false)
+{}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// General machine state
+//
+
+void tandberg_tdv2100_disp_logic_device::device_start()
+{
+	m_beep_trigger = timer_alloc(FUNC(tandberg_tdv2100_disp_logic_device::end_beep), this);
+	m_speed_ctrl = timer_alloc(FUNC(tandberg_tdv2100_disp_logic_device::expire_speed_check), this);
+
+	save_item(NAME(m_page_roll));
+	save_item(NAME(m_cursor_row));
+	save_item(NAME(m_cursor_col));
+	save_item(NAME(m_cursor_row_input));
+	save_item(NAME(m_cursor_col_input));
+	save_item(NAME(m_underline_input));
+	save_item(NAME(m_video_enable));
+	save_item(NAME(m_speed_check));
+	save_item(NAME(m_data_terminal_ready));
+	save_item(NAME(m_request_to_send));
+	save_item(NAME(m_rx_handshake));
+	save_item(NAME(m_tx_handshake));
+}
+
+void tandberg_tdv2100_disp_logic_device::device_reset()
+{
+	m_cursor_row_input = false;
+	m_cursor_col_input = false;
+	m_page_roll = 0;
+	char_to_display(0x19);
+	m_frame_counter = 0;
+	m_underline_input = false;
+	m_video_enable = true;
+	m_data_terminal_ready = false;
+	m_request_to_send = false;
+	m_data_pending_kbd = false;
+	m_uart->write_xr(1);
+	m_uart->write_xr(0);
+	w_cleark(1);
+	w_cleark(0);
+
+	bool force_on_line = m_dsw_u39->read()&0x020;
+	set_uart_state_from_switches();
+	if(force_on_line)
+	{
+		w_linek(1);
+		w_linek(0);
+		w_transk(1);
+		w_transk(0);
+	}
+	update_rs232_lamps();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Graphics and character data
+//
+
+void tandberg_tdv2100_disp_logic_device::char_to_display(uint8_t byte)
+{
+	// Note: There are no checks in hardware to catch an invalid cursor-position.
+	//       To clear an invalid cursor placement, the cursor-movement mechanism
+	//       may eventually catch it later, but it can also be manually sorted
+	//       out by pressing the CR and/or HOME key.
+	if(m_cursor_row_input)
+	{
+		place_cursor(byte, m_cursor_col);
+		m_cursor_col_input = true;
+		m_cursor_row_input = false;
+	}
+	else if(m_cursor_col_input)
+	{
+		place_cursor(m_cursor_row, byte);
+		m_cursor_col_input = false;
+	}
+	else
+	{
+		// Ordinary input
+		if((byte&0x7f) < 0x20 || (byte&0x7f) == 0x7f)
+		{
+			// Control characters
+			bool auto_roll_up = !(m_dsw_u39->read()&0x100);
+			bool clear_ascii_hs_by_cleark = ((m_dsw_u39->read()&0x003) == 0x001);
```

Comment:
You still aren’t thinking about avoiding unnecessary I/O port reads.  You’re reading the same port twice in a row here.

Reply:
Adjusted.